mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageDAO.java

```diff
@@ -0,0 +1,322 @@
+                .from(TABLE_NAME)
+                .where(eq(MESSAGE_ID, bindMarker(MESSAGE_ID))));
+    }
+
+    public CompletableFuture<Void> save(Mailbox mailbox, MailboxMessage message) throws MailboxException {
+        try {
+            CassandraMessageId messageId = (CassandraMessageId) message.getMessageId();
+            BoundStatement boundStatement = insert.bind()
+                .setUUID(MESSAGE_ID, messageId.get())
+                .setDate(INTERNAL_DATE, message.getInternalDate())
+                .setInt(BODY_START_OCTET, (int) (message.getFullContentOctets() - message.getBodyOctets()))
+                .setLong(FULL_CONTENT_OCTETS, message.getFullContentOctets())
+                .setLong(BODY_OCTECTS, message.getBodyOctets())
+                .setBytes(BODY_CONTENT, toByteBuffer(message.getBodyContent()))
+                .setBytes(HEADER_CONTENT, toByteBuffer(message.getHeaderContent()))
+                .setList(PROPERTIES, message.getProperties().stream()
+                    .map(x -> typesProvider.getDefinedUserType(PROPERTIES)
+                        .newValue()
+                        .setString(Properties.NAMESPACE, x.getNamespace())
+                        .setString(Properties.NAME, x.getLocalName())
+                        .setString(Properties.VALUE, x.getValue()))
+                    .collect(Collectors.toList()))
+                .setList(ATTACHMENTS, message.getAttachments().stream()
+                    .map(this::toUDT)
+                    .collect(Collectors.toList()));
+
+            return cassandraAsyncExecutor.executeVoid(setTextualLineCount(boundStatement, message.getTextualLineCount()));
+
+        } catch (IOException e) {
+            throw new MailboxException("Error saving mail", e);
+        }
+    }
+
+    private BoundStatement setTextualLineCount(BoundStatement boundStatement, Long textualLineCount) {
+        return Optional.ofNullable(textualLineCount)
+               .map(value -> boundStatement.setLong(TEXTUAL_LINE_COUNT, value))
+               .orElseGet(() -> boundStatement.setToNull(TEXTUAL_LINE_COUNT));
+    }
+
+    private UDTValue toUDT(MessageAttachment messageAttachment) {
+        return typesProvider.getDefinedUserType(ATTACHMENTS)
+            .newValue()
+            .setString(Attachments.ID, messageAttachment.getAttachmentId().getId())
+            .setString(Attachments.NAME, messageAttachment.getName().orNull())
+            .setString(Attachments.CID, messageAttachment.getCid().transform(Cid::getValue).orNull())
+            .setBool(Attachments.IS_INLINE, messageAttachment.isInline());
+    }
+
+    private ByteBuffer toByteBuffer(InputStream stream) throws IOException {
+        return ByteBuffer.wrap(ByteStreams.toByteArray(stream));
+    }
+    
+    public CompletableFuture<Stream<Pair<MailboxMessage, Stream<MessageAttachmentById>>>> retrieveMessages(List<ComposedMessageIdWithMetaData> messageIds, FetchType fetchType, Optional<Integer> limit) {
```

Comment:
You should not build MailboxMessage if the table you query doesn't the information needed for that, it's a design issue IMO.